File: graph/model/storage.go
package model

import "time"

type StoragePath struct {
	StorageID     string    `json:"storageId"`
	Urls          string    `json:"urls"`
	Weight        int       `json:"weight"`
	MaxStorage    int       `json:"maxStorage"`
	CanSeal       bool      `json:"canSeal"`
	CanStore      bool      `json:"canStore"`
	Groups        *string   `json:"groups"`
	AllowTo       *string   `json:"allowTo"`
	AllowTypes    *string   `json:"allowTypes"`
	DenyTypes     *string   `json:"denyTypes"`
	Capacity      int       `json:"capacity"`
	Available     int       `json:"available"`
	FsAvailable   int       `json:"fsAvailable"`
	Reserved      int       `json:"reserved"`
	Used          int       `json:"used"`
	LastHeartbeat time.Time `json:"lastHeartbeat"`
	HeartbeatErr  *string   `json:"heartbeatErr"`
	AllowMiners   string    `json:"allowMiners"`
	DenyMiners    string    `json:"denyMiners"`
}

File: ui/src/gql/storage.ts
import gql from 'graphql-tag'

export const storagePathFragment = gql`
  fragment StoragePathAll on StoragePath {
    id
    storageId
    urls
    weight
    maxStorage
    canSeal
    canStore
    type
    groups
    allowTo
    allowTypes
    denyTypes
    capacity
    available
    fsAvailable
    reserved
    used
    lastHeartbeat
    heartbeatErr
    allowMiners
    denyMiners
  }`

export const storageLivenessFragment = gql`
  fragment StorageLivenessAll on StorageLiveness {
    storageId
    url
    lastChecked
    lastLive
    lastDead
    lastDeadReason
  }`

export const GetStorageStats = gql`
  query GetStorageStats {
    storageStats {
      type
      totalAvailable
      totalCapacity
      totalFsAvailable
      totalReserved
      totalUsed
    }
  }`

export const GetStorage = gql`
  query GetStorage($id: String!) {
    storage(id: $id) {
      id
      path {
        ...StoragePathAll
      }
      liveness {
        ...StorageLivenessAll
      }
    }
  }
  ${storagePathFragment}
  ${storageLivenessFragment}
`

export const GetStorages = gql`
    query GetStorages {
      storages {
        id
        path {
          ...StoragePathAll
        }
        liveness {
          ...StorageLivenessAll
        }
      }
    }
  ${storagePathFragment}
  ${storageLivenessFragment}
`

export const GetStoragePaths = gql`
    query GetStoragePaths {
      storages {
        id
        path {
          ...StoragePathAll
        }
      }
    }
  ${storagePathFragment}
`
